libs/libc/stdio/lib_fopencookie.c

```diff
@@ -0,0 +1,229 @@
+
+  ret = cookie->cookie_io.read(cookie->cookie, buf, nbytes);
+  return ret;
+}
+
+static ssize_t fwritecookie(FAR struct file *filep, FAR const char *buf,
+                            size_t nbytes)
+{
+  struct cookie_s *cookie = (struct cookie_s *)filep->f_priv;
+  int ret;
+
+  DEBUGASSERT(cookie->cookie != NULL && cookie->cookie_io.write != NULL);
+
+  ret = cookie->cookie_io.write(cookie->cookie, (FAR const char *)buf,
+                               nbytes);
+  return ret;
+}
+
+static off_t fseekcookie(FAR struct file *filep, off_t offset, int whence)
+{
+  struct cookie_s *cookie = (struct cookie_s *)filep->f_priv;
+  int ret;
+
+  DEBUGASSERT(cookie->cookie != NULL && cookie->cookie_io.seek != NULL);
+
+  ret = cookie->cookie_io.seek(cookie->cookie, &offset, whence);
+  return ret;
+}
+
+static int fclosecookie(FAR struct file *filep)
+{
+  struct cookie_s *cookie = (struct cookie_s *)filep->f_priv;
+  int ret;
+
+  DEBUGASSERT(cookie->cookie != NULL && cookie->cookie_io.close != NULL);
+
+  ret = cookie->cookie_io.close(cookie->cookie);
+
+  free(filep->f_priv);
+  free(filep->f_inode);
+  close(cookie->cookie_fd);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: fopencookie
+ ****************************************************************************/
+
+FAR FILE *fopencookie(void *restrict cookie, const char *restrict mode,
+                      cookie_io_functions_t io_funcs)
+{
+  FAR FILE *filestructp = NULL;
+  FAR struct file filep;
+  FAR struct inode *inode;
+  FAR struct cookie_s *cookie_priv;
+  int oflags;
+  int ret;
+  int fd;
+
+  /* Map the open mode string to open flags */
+
+  oflags = lib_mode2oflags(mode);
+  if (oflags < 0)
+    {
+      return NULL;
+    }
+
+  /* Allocate necessary structures. */
+
+  inode = lib_zalloc(sizeof(struct inode));
+  if (inode == NULL)
+    {
+      set_errno(-ENOMEM);
+      goto fopencookie_return;
+    }
+
+  cookie_priv = lib_zalloc(sizeof(struct cookie_s));
+  if (cookie_priv == NULL)
+    {
+      free(inode);
+      set_errno(-ENOMEM);
+      goto fopencookie_return;
+    }
+
+  memset(&filep, 0, sizeof(filep));
+
+  /* Fill cookie_priv structure. We need to add cookie provided by user
+   * and functions providd by user.
+   */
+
+  cookie_priv->cookie = cookie;
+  cookie_priv->cookie_io.read  = io_funcs.read;
+  cookie_priv->cookie_io.write = io_funcs.write;
+  cookie_priv->cookie_io.seek  = io_funcs.seek;
+  cookie_priv->cookie_io.close = io_funcs.close;
+
+  /* Add file operations */
+
+  inode->u.i_ops = &cookie_fops;
+
+  filep.f_oflags = oflags;
+  filep.f_inode = inode;
+  filep.f_priv = cookie_priv;
+
+  /* And get file descriptor. */
+
+  fd = file_allocate_from_tcb(nxsched_self(), filep.f_inode, filep.f_oflags,
```

Comment:
libc can't call os internal function directly

Reply:
If I create `fs/vfs/fs_fopencookie.c` with function `fs_fopencookie` that would call the internal function (similarly to `fs_fdopen`), would that be ok?

Reply:
I think we don't need allocate fd here, it's enough to set fs_fd to -1, just like how musl implement fopencookie:
https://github.com/bminor/musl/blob/master/src/stdio/fopencookie.c#L121

Reply:
But I need valid fd to be able to call proper functions, no? `fopencookie` returns `FILE` structure which contains file descriptor. Then if IO operation (lets say `write`) is called, NuttX uses the file descriptor to get `struct file` which contains pointer to `inode` which has `u.i_ops` with IO operations. And this operations lead to private functions in `lib_fopencookie.c`. 

MUSL implementation is a bit different, they have pointers to IO directly in `FILE`, we don´t. So I think we always need file descriptor if we want to get there from `FILE` stream.

Reply:
Ok, you implementation reuse file_operation_s infrastructure, but it mayn't work in protected/kernel mode, since fclosecookie... is part of kernel, but call the callback provided by userspace directly.

Reply:
> Ok, you implementation reuse file_operation_s infrastructure, but it mayn't work in protected/kernel mode, since fclosecookie... is part of kernel, but call the callback provided by userspace directly. I would suggest that you have to add these callback to file_struct and call them from libc/stdio directly.
> 
> @patacongo what are you thinking?

_[I have not looked at this particular PR, but I can give my usual response when callbacks from the OS are suggested]_

Direct callbacks from the OS into user code have always been forbidden.  I don't think we should permit that now.  Some justifications:

- This is not the POSIX way to do things.  There is no way to do such things under any POSIX OS.  We have always been concerned about POSIX compliant and portability.  Linux code is often ported to NuttX. NuttX code should be runnable on any any POSIX OS as well (with the appropriate drivers).  To accomplish these goals, callbacks must not be used.

- The callback would probably be running on an kernel stack and not the user stack.  In that case, TLS data on the stack would not be accessible.  The callback were asynchronous with the tasking.  A callback is not a context switch; the interrupt user task is still in place and the OS thinks that is that task that is running.  So any OS interactions in this bastardized state could likely be fatal.

- The callback would run at whatever the priority of the calling logic is running at.  This will break real time behavior which must be strict priority-based scheduling.

There are several differ build modes for NuttX:  The FLAT build, the PROTECTED build for better security, and the KERNEL build for full support of processes much like Linux.  In the PROTECTED and KERNEL builds, there are more significant technical issues.

There is some documentation of at least the PROTECTED build here:  https://cwiki.apache.org/confluence/display/NUTTX/NuttX+Protected+Build

- In the PROTECTED and KERNEL build modes, the kernel runs in supervisor mode in a supervisor address space while all user applications run in non-privileged user mode in a user address space.  In the PROTECTED mode, you _could_ call from the kernel supervisor code into user code but this would, however, be a gaping security hole and, hence, must never be done.  Potentially hostile user-supplied code must never run in supervisor mode in this build.

- The KERNEL build differs in that each _process_ has its own address environment and, in this build, the system will crash if such a thing is attempted.  That is because the notification arrives asynchronously with tasking; you never know which process is running and you would call into the wrong address space.  This could be fixed be changing the address environment prior to each callback, that would have big performance and complexity issues.  It is better to do things in the POSIX way.

So even if you use only the FLAT build, direct callbacks from the OS into the user code must be prohibited.  What is the POSIX way of doing an aynchronous notification?  Signals.  In a POSIX OS, you should use standard POSIX signaling and avoid even thinking about callbacks.

Drivers under drivers/ that do asynchronous notifications use signals for this purpose.  Consider drivers/input/button_upper.c.  This is a simple button driver that sends a notification to a user task when a button is pressed or release:

- Look at the function btn_ioctl().  The BTNIOC_REGISTER command is used to register a notification.  The process ID of the user thread is provided and retained for the notification.

- When the button press event occurs, the notification signal is sent in btn_sample().

- A signal handler in the user thread that registered for the notification can then run, in the correct task context, in user mode, and at the correct priority assigned to the task.

Reply:
Since stdio is a pure usespace library, I would suggest keep all fopencookie related stuff in libc to avoid the complexity of kernel/userspace interactions.
@michallenc you can achieve this by:

1. Add cookie_io_functions_t to file_struct(FILE)
2. Assign the internal callback which read/write/seek through fd to cookie_io_functions_t in fopen
3. Assign caller supplied callback in fopencookie
4. Replace read/write/seek with cookie_io_functions_t's one

Reply:
> Since stdio is a pure usespace library, I would suggest keep all fopencookie related stuff in libc to avoid the complexity of kernel/userspace interactions.

That would not solve the technical problem.  It eliminates the kernal/user privilege issue, but is otherwise insufficient.  That is because the caller executes in Task/Process A but the user-implementation of the file is implemented in a different Task/Process B.  There are lots of problems with that:

- You would be executing code in Task B in the context of Task A.
- That means Task B would use Task A's stack and file structures, not those of Task B. 
- The Task B file system code could not access Task B's TLS data .
- From the standpoint of the OS, Task A is running, not Task B.  Task B could not access the OS as Task B and many things just would not work properly.
- The logic in Task B would run with with priority and identity of Task A.
- It would not work at all in the KERNEL build mode because Task B resides in a different process address space than does Task A.  That would crash immediately and dramatically.
- etc. Many more issues.

I don't believe that any of that behavior is acceptable.

I still believe that the only viable solution must depend on using a proper POSIX IPC like signals, message queues, fifos, local networks, etc.  Then Task B runs in the proper context and address space of Task B.

Reply:
FILE is private to a specific task/process, it's illegal to share FILE * between the different tasks. So, @patacongo the scenario you describe can never happen.

Reply:
> Add cookie_io_functions_t to file_struct(FILE)
Assign the internal callback which read/write/seek through fd to cookie_io_functions_t in fopen
Assign caller supplied callback in fopencookie
Replace read/write/seek with cookie_io_functions_t's one

I am not sure this would solve the problem. Because the callback to `fopencookie` would still go through `fs/vfs`. Now this would work in FLAT build but probably not in protected build. What we would need is that `libc` would call the callbacks directly for `FILE *` if custom streams are defined as it is done in ucLibc for example https://elixir.bootlin.com/uclibc-ng/v1.0.7/source/libc/stdio/fread.c#L68 This way we would not need  But that is too big change now.

Reply:
> > Add cookie_io_functions_t to file_struct(FILE)
> > Assign the internal callback which read/write/seek through fd to cookie_io_functions_t in fopen
> > Assign caller supplied callback in fopencookie
> > Replace read/write/seek with cookie_io_functions_t's one
> 
> I am not sure this would solve the problem. Because the callback to `fopencookie` would still go through `fs/vfs`.

In my suggestion, you should stop to implement file_operation_s and call file_allocate_from_tcb. How do you go through fs/vfs?

> Now this would work in FLAT build but probably not in protected build. What we would need is that `libc` would call the callbacks directly for `FILE *` if custom streams are defined as it is done in ucLibc for example https://elixir.bootlin.com/uclibc-ng/v1.0.7/source/libc/stdio/fread.c#L68 This way we would not need But that is too big change now.

This is the right way to go. And most change is minor.

Reply:
> In my suggestion, you should stop to implement file_operation_s and call file_allocate_from_tcb. How do you go through fs/vfs without them?

Yes, but you mention 

> Assign the internal callback which read/write/seek through fd to cookie_io_functions_t in fopen

How would I get fd without actually using `file_allocate_from_tcb`? I cannot even call `fopen` without that.

Reply:
fd shouldn't reference in common code, but in the callback provided by stdio self. Actually fd should change to void * instead.